[FLINK-19662][runtime] Exposing the JobManager's memory configuration through /jobmanager/config

XComp opened the pull request:
## What is the purpose of the change

We want to expose the JobManager's effective configuration through the `/jobmanager/config` REST endpoint. That way, we can expose these values in the web UI as it is suggested in [FLIP-104](https://cwiki.apache.org/confluence/display/FLINK/FLIP-104%3A+Add+More+Metrics+to+Jobmanager).

## Brief change log

- Added method to generate dynamic parameters from configuration
- Added dynamic parameters to BashJavaUtils' JobManager argument extraction and integrated the new functionality in multiple script starting the JobManager directly or indirectly.

## Verifying this change

- Added test to JobManagerProcessUtilsTest to check whether the dynamic parameter list generation works correctly.
- Extended BashJavaUtilsITCase to check whether the dynamic properties are returned by BashJavaUtilsITCase.
This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not documented (the specific configuration parameters exposed through the REST endpoint are not documented in general)

XComp commented:
@tillrohrmann Feel free to have a look.

XComp commented:
I created [FLINK-19788](https://issues.apache.org/jira/browse/FLINK-19788) as ShuffleCompressionITCase.testDataCompressionForBlockingShuffle [failed](https://dev.azure.com/mapohl/flink/_build/results?buildId=92&view=logs&j=70ad9b63-500e-5dc9-5a3c-b60356162d7e&t=944c7023-8984-5aa2-b5f8-54922bd90d3a) even though it seems to be unrelated to my changes.

tillrohrmann commented on flink-dist/src/main/flink-bin/bin/jobmanager.sh:
```diff
@@ -51,6 +51,10 @@ if [[ $STARTSTOP == "start" ]] || [[ $STARTSTOP == "start-foreground" ]]; then
     fi
+
+    if [ ! -z "${DYNAMIC_PARAMETERS+x}" ]; then
+        for v in ${DYNAMIC_PARAMETERS[@]}; do args+=($v); done
```
Why do we have to add it in a loop to `args`? Is this some bash magic?

XComp replied:
When I implemented it firstly, I ran into issues. Hence, I used the for loop option. I addressed it again and it worked. No clue, what was wrong in the first place. 🤔 Anyway, I fixed it. Now, the `args` array simply gets concatenated with the `DYNAMIC_PARAMETERS` array.

tillrohrmann commented on flink-dist/src/main/flink-bin/bin/jobmanager.sh:
```diff
@@ -51,6 +51,10 @@ if [[ $STARTSTOP == "start" ]] || [[ $STARTSTOP == "start-foreground" ]]; then
         args+=("${WEBUIPORT}")
     fi
+
+    if [ ! -z "${DYNAMIC_PARAMETERS+x}" ]; then
```
Why is a `+x` appended to `DYNAMIC_PARAMETERS`? In the other scripts this suffix is missing.

XComp replied:
True I forgot to align all of that. The variable extension is not necessary in our case. Hence, I removed it again aligning all three code snippets.

tillrohrmann commented on flink-dist/src/test/java/org/apache/flink/dist/BashJavaUtilsITCase.java:
```diff
@@ -72,20 +79,23 @@ public void testGetTmResourceParamsConfigsWithDynamicProperties() throws Excepti
 	@Test
 	public void testGetJmResourceParams() throws Exception {
-		int expectedResultLines = 1;
+		int expectedResultLines = 2;
```
I guess that this commit should be merged together with the previous one because w/o, the previous commit should fail in this test here.

XComp replied:
Yup, I planned to squash all the commits in a proper way after the review is done.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @XComp. I think your changes look good for the standalone mode and when using the entrypoint scripts. But I think they are missing the programmatic cluster creation for Yarn and K8s. I think we should also add the dynamic properties in `JavaCmdJobManagerDecorator.getJobManagerStartCommand` and `YarnClusterDescriptor.setupApplicationMasterContainer`.

XComp commented:
@tillrohrmann I addressed your issues and waited for the Azure pipeline to succeed. I continued with reorganizing the commits. Feel free to have another look. In the meantime, I'm still trying to test the changes manually.

XComp commented:
@tillrohrmann I tested whether the memory configuration pops up when accessing `job-managers/config` manually for:
* local standalone cluster (using `./bin/jobmanager.sh` through `./bin/start-cluster.sh`)
* Kubernetes Session Cluster (using `./bin/kubernetes-session.sh # ...`)
* Kubernetes Job Cluster (using `./bin/flink run-application -p 8 -t kubernetes-application # ...`)
* Yarn Session Cluster (using `./bin/yarn-session.sh -jm 1024m -tm 4096`)
* Yarn Job Cluster (using `./bin/flink run -m yarn-cluster -p 4 # ...`)

Additionally, I verified the parameters passing for `./bin/mesos-appmaster.sh` and `./bin/mesos-appmaster-job.sh` which both execute `./bin/mesos-jobmanager.sh`. I also did a code review of passing the configuration trough `MesosJobClusterEntrypoint` and `MesosSessionClusterEntrypoint` to `LaunchableMesosWorker`.

All looks good. So, I guess, the code is ready to get another review.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/DynamicParametersConfigurationParserFactory.java:
```diff
@@ -0,0 +1,57 @@
+import static org.apache.flink.runtime.entrypoint.parser.CommandLineOptions.DYNAMIC_PROPERTY_OPTION;
+
+/**
+ * {@code DynamicParametersCOnfigurationParserFactory} can be used to extract the dynamic parameters
```
typo: COnfiguration

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnEntrypointUtils.java:
```diff
@@ -51,8 +56,24 @@
+	private static final Logger LOG = LoggerFactory.getLogger(YarnEntrypointUtils.class);
+
+	public static Configuration loadConfiguration(String workingDirectory, String[] args, Map<String, String> env) {
+		Configuration dynamicParameters = null;
+		final CommandLineParser<Configuration> commandLineParser = new CommandLineParser<>(
+				new DynamicParametersConfigurationParserFactory());
+
+		try {
+			dynamicParameters = commandLineParser.parse(args);
+		} catch (FlinkParseException e) {
+			LOG.error("Could not parse command line arguments {}.", args, e);
+			commandLineParser.printHelp(StandaloneSessionClusterEntrypoint.class.getSimpleName());
+			System.exit(1);
+		}
```
This snipped looks quite similar to what you've put into the `KubernetesEntrypointUtils`. I'd suggest to factor it out and to reuse it. Maybe even pass `dynamicParameters` into this method instead of `args`.

XComp replied:
You have a good point. I looked into the code again and realized that it's not ideal (the current version and your proposal moving everything in a single utility method). I don't consider calling `System.exit(..)` within a utility method good coding. Such a call should be located in the main method. 

Moving the left-over code into a utility method would have meant splitting up the error handling (since `CommandLineParser.printHelp(..)` would have still been called within the utility method).

I didn't consider splitting up the error handling being a good approach either. Hence, I moved the dynamic parameter parsing into each main method in [7f722c9](https://github.com/apache/flink/pull/13747/commits/7f722c99b83591793147b2dda952a3acadf5f2b7). What's your take on this?

FYI: The commit above also fixed a bug that used the wrong class in `CommandLineParser.printHelp(..)`.

tillrohrmann reviewed (approved):
Thanks a lot for updating this PR @XComp. The changes look good to me. I had one minor and one normal comment about deduplicating code. Please take a look and let me know once you've updated it. I will then merge this PR.

tillrohrmann commented on flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnSessionClusterEntrypoint.java:
```diff
@@ -73,7 +76,17 @@ public static void main(String[] args) {
 		}
 
-		Configuration configuration = YarnEntrypointUtils.loadConfiguration(workingDirectory, env);
+		final CommandLineParser<Configuration> commandLineParser = new CommandLineParser<>(new DynamicParametersConfigurationParserFactory());
+		Configuration dynamicParameters = null;
+		try {
+			dynamicParameters = commandLineParser.parse(args);
+		} catch (FlinkParseException e) {
+			LOG.error("Could not parse command line arguments {}.", args, e);
+			commandLineParser.printHelp(YarnSessionClusterEntrypoint.class.getSimpleName());
+			System.exit(1);
+		}
```
Maybe one could introduce

```
private static <T> T parseArgumentsOrExit(
			Logger logger,
			String componentName,
			String[] args,
			ParserResultFactory<T> parserResultFactory) {

		final CommandLineParser<T> commandLineParser = new CommandLineParser<>(parserResultFactory);
		try {
			return commandLineParser.parse(args);
		} catch (FlinkParseException e) {
			logger.error("Could not parse command line arguments {}.", args, e);
			commandLineParser.printHelp(componentName);
			System.exit(1);
			return null;
		}
	}
```

into `ClusterEntrypointUtils` and use it across the cluster entrypoints. This would also ensure that we fail with a consistent error code if the parsing fails.

XComp replied:
Ok, fair enough. I addressed the comments and pushed the changes. FYI: I also moved `ClusterEntrypointUtils` into the package of `ClusterEntrypoint` (i.e. `org.apache.flink.runtime.entrypoint`) since I wanted to use `ClusterEntrypoint.STARTUP_FAILURE_RETURN_CODE` in the newly introduced utility method.

Alternatively, I could have made `ClusterEntrypoint.STARTUP_FAILURE_RETURN_CODE` public. But moving `ClusterEntrypointUtils` felt like the right thing to do considering that other utility classes (e.g. `ClusterEntrypointExceptionUtils`) are also located in `org.apache.flink.runtime.entrypoint`. Any objections against that?

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @XComp. I think there is compilation error in `YarnApplicationClusterEntryPoint` which we need to fix.

tillrohrmann reviewed (approved):
I think the compilation error should be fixed. I will merge this PR once AZP gives green light.
